### DISCUSSION TITLE
change return type of useQuery to discriminating union

### DISCUSSION THREAD

**Discussion by MrLoh** [TanStack/query#1102] [General]
To fully take advantage of typescript, it would be nice if the return type of `useQuery` would be a [discriminating union](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions). 

```ts
const Component = () => {
  const res = useQuery<Data, Error>('key', fetchData);
  return res.status === 'loading' ? (
    <div>Loading ...</div>
  ) : res.status === 'error' ? (
    <div>{res.error.message}</div> // ts understands that error is not undefined
  ) : res.status === 'success' ? (
    <div>{res.data.name}</div> // ts understands that data is not undefined
  ) : null;
};
```

Has this been discussed, are there any concerns about this? I could potentially look into providing a PR.

This probably applies to the other hooks too.

**Comment by TkDodo** [comment]
I think this would be awesome 😎 , not sure if there is a reason to _not_ have the status property discriminate the data / error fields. It would give me another argument to use the status field over loading booleans :)

**Comment by TkDodo** [comment]
@boschni pointed out some time ago that the discriminating unions have been removed:

https://github.com/tannerlinsley/react-query/discussions/922#discussioncomment-56912

**Comment by MrLoh** [comment]
I don't understand the argument though. I created a PR to add them back, #1108 . Maybe we can continue the discussion there.
